### ISSUE TITLE
chore(ios): 'run script' settings in Xcode configs should refer to /usr/bin/env bash, not bin/sh

### ISSUE THREAD

**Issue by jahorton** [keymanapp/keyman#9003]
Pretty much the title.

I didn't catch this at the time, but it turns out that #8260's changes can cause issues when building within Xcode; the default bin/sh on my Mac dev machine predates the availability of `declare -g` as introduced in that PR here:

https://github.com/keymanapp/keyman/blob/836d751090e467aaa61dad438f901404fdf45128/resources/build/build-utils.sh#L355-L356

Due to how we include `xcode-utils.sh` in the Run Script's script, its shebang can get bypassed.  It turns out that the code _was_ being run from within _**a**_ bash - but an older version than returned by `where bash` - even in the same script!

Changing the "Run Script" shell element to read `/usr/bin/env bash` instead of `bin/sh` allowed me to work around that problem; Xcode then selected the "right" version of bash, where `declare -g` was valid, and the build succeeded.

If there are equivalent steps for our Mac builds, we should probably double-check their corresponding settings.

**Comment by mcdurdin** [comment]
I'm going to mark this Future as a low-level task we can undertake if/when we encounter again, but agree longer term should clean this up. Probably a good one to do alongside build.sh rewrites for those platforms.

**Comment by darcywong00** [comment]
Jotting some notes from Slack:

@jahorton about building iOS on Apple Silicon with Xcode
> For developing on a Silicon Mac, it's highly advisable to make a symlink **from** /usr/local/bin **to** /opt/homebrew/bin.  Adding a `-l` option on each "Shell: " entry (for Xcode run script phases) then puts the symlink accessible via path.
>
>That's enough for a `which bash` within the Xcode shell scripts to pick up the right version... within the scripts.  Just... not to actually launch the right version directly within the correct bash.  :(
For that, we then…
>* create a specialized shell script we can call that's fully formed, launch that within a sub-shell `bash` instance.
>    * This looks to be "enough" to resolve the issue.
>* If it doesn't seem to work immediately, try a "clean build folder" operation; that certainly helped me!

----
@mcdurdin 
> OK, so if we move the scripts out of the xcodeproj as we've been talking about doing anyway, then we can use the usr/bin/env shebang on those scripts, which seems like a more maintainable long-term solution too, right?

**Comment by mcdurdin** [comment]
Please note, I don't think we should be doing the steps that @jahorton suggested above.

It should be enough to follow the notes already described in this issue -- either use the `#!/usr/bin/env bash` shebang in any xcodeproj build steps, or better yet, move any script content out into separate bash scripts that are just called from the xcodeproj, because they are far more maintainable.

**Comment by jahorton** [comment]
> Please note, I don't think we should be doing the steps that @jahorton suggested above.

Please note, I wouldn't have suggested them if they weren't necessary.

Xcode launches with a sanitized set of environment variables - which includes a sanitized $PATH.  Without running code that may be run per-launch, Xcode on Apple Silicon _cannot_ launch the Homebrew bash version from IDE-triggered builds.  That's what the proposed changes are there to work around, so that we can trigger the right `bash` from within the Xcode-sanitized one.

**Comment by mcdurdin** [comment]
Yeah okay, I see what you mean. But I am still not excited about having to use a symlink to resolve this as that is also fragile, a big system-wide hammer that may lead to other conflicts.

I wonder if we could resolve by a wrapper script in the repo which has enough knowledge about the different environments we are aware of. Unlike all our other scripts, it would be explicitly designed to run in any version of bash available on mac (e.g. we'd avoid pathing fixups etc). Then if there is an upgrade to homebrew or macos that changes paths around, it's just a single place to fix in our repo which is easy enough to maintain. Seems like less work?

Put this in /resources/build/xcode-wrap.sh.

Then each script in the .xcodeproject would simply be:
```
.../resources/build/xcode-wrap.sh .../mac/my-file.sh [params...]
```

pseudocode

```sh
#!/bin/bash

if intel-silicon && intel-homebrew-bash-exists; then
  /.../intel-silicon-homebrew-bash/bash "$@" || exit $?
elif arm-silicon && arm-homebrew-bash-exists; then
  /.../arm-silicon-homebrew-bash/bash "$@" || exit $?
else
  >&2 echo "Could not start build due to missing homebrew bash"
  exit 55
fi 
```

**Comment by jahorton** [comment]
> The problem is that /usr/bin/env bash will still just pickup the system default bash because xcode has sanitized the path (un)helpfully for us.

Yep.  We can still get the system-wide unsanitized environment (including its `$PATH`) via login-mode shell (`-l`) _within the launched script_, but that `$PATH` isn't consulted by Xcode when it actually selects the shell.

**Comment by mcdurdin** [comment]
OK, in that case, am I right in thinking that the launcher script is actually dead simple -- it is just a one-liner `/usr/bin/env bash "$@"`?

**Comment by jahorton** [comment]
Assuming you put `/opt/homebrew/bin` in your system's `$PATH` in some form, be it via the symlink I suggested or just as an explicit `$PATH` entry, that should be fine.

**Comment by mcdurdin** [comment]
True. So perhaps there is still real value in xcode-wrap being knowledgeable about homebrew bash, in terms of reducing friction for external contributors.

**Comment by mcdurdin** [comment]
Note: for sample projects we can't use xcode-wrap or other external scripts in that way. They should probably be in their own repositories so that we don't leak our repo config into them, but that's a separate project.

**Comment by jahorton** [comment]
In the meantime, in case anyone else runs into the same issue - or in case I need to keep on re-referencing it - https://stackoverflow.com/a/33237548 is quite useful for forcing Xcode to ignore santization, allowing it to directly launch homebrew bash if it's in the standard environment `$PATH`.  

This must be re-run after restarting the computer, and I usually need to "clean build folder" once after running the command.  After that, builds work again.
